src/transactions/TransactionFrame.cpp

```diff
@@ -1511,6 +1522,172 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
+    }
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    LedgerTxn ltxTx(ltx);
+    auto const& resources = sorobanResources();
+    auto lcl = app.getLedgerManager().getLastClosedLedgerNum();
+    auto const& expirationSettings = app.getLedgerManager()
+                                         .getSorobanNetworkConfig(ltxTx)
+                                         .stateExpirationSettings();
+
+    auto maxExpirationLedger = lcl + expirationSettings.maxEntryExpiration;
+    auto minRestorableExpirationLedger =
+        lcl + expirationSettings.minRestorableEntryExpiration;
+    auto minTempExpirationLedger =
+        lcl + expirationSettings.minTempEntryExpiration;
+    auto autoBumpLedgers = expirationSettings.autoBumpLedgers;
+
+    // Applies the correct lifetime to LE. If lifetime has not changed from
+    // pre-tx apply value, returns false. Otherwise, returns true
+    auto bump = [&](LedgerEntry& le, bool autoBump, bool enforceMinimum) {
+        // If an entry is being created for the first time, there is no original
+        // lifetime, so set it to lcl for fee calculation purposes
+        auto iter = mOriginalLifetimes.find(LedgerEntryKey(le));
+        uint32_t preApplyExpiration =
+            iter == mOriginalLifetimes.end() ? lcl : iter->second;
+        uint32_t postApplyExpiration = getExpirationLedger(le);
+        uint32_t minimumForEntry = preApplyExpiration;
+
+        // First, calculate minimum lifetime that should be applied. This value
+        // is (preApplyLifetime + autobump) or minLifetimeForType, whichever is
+        // greater.
+        if (autoBump)
+        {
+            // Note: preApplyExpiration is guarenteed to be a valid expiration
+            // ledger so this can't overflow
+            minimumForEntry = preApplyExpiration + autoBumpLedgers;
+        }
+
+        if (enforceMinimum)
+        {
+            if (isTemporaryEntry(le.data))
+            {
+                minimumForEntry =
+                    std::max(minimumForEntry, minTempExpirationLedger);
+            }
+            else
+            {
+                minimumForEntry =
+                    std::max(minimumForEntry, minRestorableExpirationLedger);
+            }
+        }
+
+        // Check if the lifetime bump applied by the host is enough to cover the
+        // required minimum
+        uint32_t lifetimeToApply = postApplyExpiration < minimumForEntry
+                                       ? minimumForEntry
+                                       : postApplyExpiration;
+
+        // Cap at max lifetime
+        lifetimeToApply = std::min(lifetimeToApply, maxExpirationLedger);
+
+        if (lifetimeToApply == preApplyExpiration)
+        {
+            return false;
+        }
+        else
+        {
+            setExpirationLedger(le, lifetimeToApply);
+            return true;
+        }
+    };
+
+    for (auto const& key : resources.footprint.readWrite)
+    {
+        auto lte = ltxTx.load(key);
```

Comment:
We can check that the LedgerKey belongs to a Soroban data entry before loading. This comment applies to the readOnly bump code below.